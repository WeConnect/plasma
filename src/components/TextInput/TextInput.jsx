// @flow
import React, { Component, type Node } from 'react';
import cx from 'classnames';

import { getDataAttrs } from '../../dataUtils';
import type {
  ChangeEventHanlders,
  Data,
  FocusEventHandlers,
  GlobalAttributes,
  SyntheticMouseEventHandler,
} from '../../types';

import styles from './style.scss';

type Props = {|
  data?: Data,
  disabled?: boolean,
  error?: boolean,
  maxLength?: number | string,
  name?: string,
  placeholder?: string,
  prefix?: Node,
  suffix?: Node,
  type?: 'text' | 'password' | 'email' | 'url',
  value?: string,
  onClick?: SyntheticMouseEventHandler<HTMLInputElement>,
  ...FocusEventHandlers<HTMLInputElement>,
  ...ChangeEventHanlders<HTMLInputElement>,
  ...GlobalAttributes,
|};

class TextInput extends Component<Props> {
  static defaultProps = {
    placeholder: 'Type something...',
  };

  render() {
    const { className, disabled, error, data, prefix, suffix, type, ...rest } = this.props;

    const inputType = type || 'text';

    const classes = cx(styles.wrapper, className, {
      [styles.wrapperDisabled]: disabled,
      [styles.wrapperError]: error,
    });

    return (
      <div {...getDataAttrs(data)} className={classes}>
        {prefix && <div className={styles.prefix}>{prefix}</div>}
        <input className={styles.input} disabled={disabled} type={inputType} {...rest} />
        {suffix && <div className={styles.suffix}>{suffix}</div>}
      </div>
    );
  }
}

TextInput.displayName = 'Plasma@TextInput';

export default TextInput;
