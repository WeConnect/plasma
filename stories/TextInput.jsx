import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import TextInput from '../src/components/TextInput/TextInput';
import withCounter from '../src/decorators/withCounter/withCounter';
import StoryState from '../StoryState';

const TextInputWithCounter = withCounter(TextInput);

const stories = storiesOf('TextInput', module);

stories.add('default', () => <TextInput onChange={action('changed')} />);

stories.add('custom placeholder', () => (
  <TextInput placeholder="Custom placeholder..." onChange={action('changed')} />
));

stories.add('prefix', () => (
  <TextInput onChange={action('changed')} prefix="$" />
));

stories.add('suffix', () => (
  <TextInput onChange={action('changed')} suffix=".com" />
));

stories.add('disabled placeholder', () => (
  <TextInput disabled onChange={action('changed')} />
));

stories.add('disabled filled', () => (
  <TextInput disabled onChange={action('changed')} value="Queen of hearts" />
));

stories.add('disabled prefix', () => (
  <TextInput disabled prefix="$" onChange={action('changed')} value="Queen of hearts" />
));

stories.add('disabled suffix', () => (
  <TextInput disabled suffix=".com" onChange={action('changed')} value="Queen of hearts" />
));

stories.add('error', () => <TextInput error onChange={action('changed')} />);

stories.add('suffix error', () => (
  <TextInput suffix=".com" error onChange={action('changed')} />
));

stories.add('prefix error', () => (
  <TextInput prefix="$" error onChange={action('changed')} />
));

stories.add('large', () => (
  <TextInput size="large" onChange={action('changed')} />
));

stories.add('large prefix', () => (
  <TextInput prefix="http://" size="large" onChange={action('changed')} />
));

stories.add('clipping parent', () => (
  <div>
    <div style={{ overflow: 'hidden', width: 100, height: 100 }}>
      <TextInput />
    </div>
  </div>
));

stories.add('max length', () => (
  <TextInput maxLength="10" onChange={action('changed')} />
));

stories.add('with counter', () => (
  <StoryState
    render={(state, setState) => (
      <TextInputWithCounter
        placeholder="Should show a counter"
        onChange={e => setState({ text: e.target.value })}
        value={state.text}
      />
    )}
  />
));

stories.add('with counter and maxLength', () => (
  <StoryState
    render={(state, setState) => (
      <TextInputWithCounter
        placeholder="Should show a counter and have max length"
        maxLength="20"
        onChange={e => setState({ text: e.target.value })}
        value={state.text}
      />
    )}
  />
));
