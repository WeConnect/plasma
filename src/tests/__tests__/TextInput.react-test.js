/* eslint-disable */

import React from 'react';
import TextInput from '../../components/TextInput/TextInput';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('TextInput Component', () => {
  test('render', () => {
    const component = renderer.create(<TextInput />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('focus', () => {
    const onFocus = jest.fn();
    const component = mount(<TextInput onFocus={onFocus} />);
    component.find('input').simulate('focus');
    expect(onFocus).toHaveBeenCalledTimes(1);
  });

  test('blur', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const component = mount(<TextInput onFocus={onFocus} onBlur={onBlur} />);
    component.find('input').simulate('blur');
    expect(onBlur).toHaveBeenCalledTimes(1);
  });
});
