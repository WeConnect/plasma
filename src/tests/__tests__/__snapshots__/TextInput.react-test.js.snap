// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component render 1`] = `
<div
  className="wrapper"
>
  <input
    className="input"
    disabled={undefined}
    onBlur={[Function]}
    onChange={undefined}
    onFocus={[Function]}
    placeholder="Type something..."
    type="text"
    value={undefined}
  />
</div>
`;
